im3components/wrf_tell/wrf_tell_counties.py

```
@@ -0,0 +1,204 @@
+        county_shapefile: str = './Geolocation/tl_2020_us_county/tl_2020_us_county.shp',
+        weight_and_mapping_file: str = './grid_cell_to_county_weight.parquet',
+        output_path: str = './County_Output_Files',
+        n_jobs: int = 1,
+) -> None:
+    """
+    Aggregate WRF output data to county level using area weighted average.
+
+    :param str wrf_file: path to the WRF output file to aggregate
+    :param list(str) wrf_variables: list of variables to aggregate
+    :param list(int) precisions: list of precisions corresponding to the variables to aggregate
+    :param str county_shapefile: path to a shapefile (.shp) with county geometries
+    :param str weight_and_mapping_file: path to read or write a weights file which maps WRF grid cell to county weight
+    :param str output_path: path to which output should be written
+    :param n_jobs: number of time slices to process in parallel
+    """
+
+    begin_time = datetime.datetime.now()
+
+    if not isfile(wrf_file):
+        print('No file to process, exiting...')
+        exit()
+
+    wrf = salem.open_wrf_dataset(wrf_file)
+
+    # if there's not already a mapping file, create one
+    if not isfile(weight_and_mapping_file):
+
+        # using the first file and time:
+        # * get the crs
+        # * create the mapping of cell index to county and weight
+        wrf_crs = wrf.pyproj_srs
+        wrf_df = wrf[wrf_variables].isel(time=0).to_dataframe().reset_index(drop=True)
+        wrf_df = gpd.GeoDataFrame(wrf_df, geometry=wrf.salem.grid.to_geometry().geometry).set_crs(wrf_crs)
+        wrf_df['cell_index'] = wrf_df.index.values
+
+        # load the counties and reproject to WRF projection
+        counties = gpd.read_file(county_shapefile)[["GEOID", "geometry"]].rename(columns={
+            'GEOID': 'FIPS',
+        }).to_crs(wrf_crs)
+
+        # find the intersection between counties and wrf cells
+        intersection = gpd.overlay(counties, wrf_df, how='intersection')
```

`how` needs to be passed an argument with a default of `intersection` incase the desired topology changes (e.g., ‘intersection’, ‘union’, ‘identity’, ‘symmetric_difference’ or ‘difference’).  

The outcome of this function should also ensure that all geometries are valid.  For example, a triangle will often be deemed invalid whether it is multipart or singlepart.  The `overlay` function does offer a `make_valid` option but it will also not always work.  If the output geodataframe has invalid geometries, I suggest logging a warning to the user with the geometry in question.

---

for `how`, i'd argue that none of the other overlay techniques would result in a valid weighted mean, so it might not be appropriate to allow topologies in this particular case -- do you think there's a use case for the other strategies?

checking for valid geometries makes sense; i'll add that in!

---

Agreed, for weighted mean - I think supporting `intersection` as you have it is the correct approach.

---

It seems that overlay will raise an error if invalid geometries are created, so i now catch that error and give a slightly better error message. I opted to still raise the error rather than proceed with warnings, since geometry problems will likely invalidate the averaging.